### PR TITLE
Fixed race condition on client reconnection logic

### LINKED ISSUE #207: Intermittent test failures in ClientErrorsTest.testConsumerReconnect
```
testConsumerReconnect(com.yahoo.pulsar.client.api.ClientErrorsTest)  Time elapsed: 5.018 sec  <<< FAILURE!
java.lang.AssertionError: expected [3] but found [2]
	at com.yahoo.pulsar.client.api.ClientErrorsTest.testConsumerReconnect(ClientErrorsTest.java:659)
```

### PR DESCRIPTION
**Pull request by merlimat**
### Motivation

There is a race condition in the reconnection logic that would make throwing an uncatched `NullPointerException` if the connection is being closed and cleaned up at a very specific point in the connection phase.

Fixes #207 

### Modifications

Double check the connection state when reacting on new connection being opened in the connection pool.

### CODE REVIEW

**Comment by merlimat** [comment]
@saandrews @rdhabalia We should consider to backport this to 1.16 branch as well. No need to release immediately 1.16.1, but to keep it there ready.

**Comment by saandrews** [comment]
👍

File: pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConnectionPool.java [hunk 1]
```diff
@@ -140,12 +141,23 @@ public void initChannel(SocketChannel ch) throws Exception {
...
+                if (log.isDebugEnabled()) {
+                    log.debug("Removing closed connection from pool: {}", v);
+                }
                 cleanupConnection(address, connectionKey, cnxFuture);
             });
 
             // We are connected to broker, but need to wait until the connect/connected handshake is
             // complete
             final ClientCnx cnx = (ClientCnx) future.channel().pipeline().get("handler");
+            if (!future.channel().isActive() || cnx == null) {
```
Comment (rdhabalia) [inline pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConnectionPool.java]: future.isSuccess() = true implies that future.channel() is active. That's guaranteed right? It should be but just wanted to confirm.

**Comment by rdhabalia** [comment]
>We should consider to backport this to 1.16 branch as well. No need to release immediately 1.16.1, but to keep it there ready

Yes, we will create a branch and will add this patch and we can wait for sometime before releasing if we want to include other patches as well.

File: pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConnectionPool.java (hunk 1 above)
Comment (merlimat) [inline pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConnectionPool.java]: The problem is that the connection was actually successfully open, thus the future itself is completed, but then immediately closed by the broker (that's what the test is reproducing).

So the future still stays set to success (future cannot change their status once they flip), but the isActive() now returns false, and the "handler" is also removed from the channel pipeline, causing the NPE when accessing cnx.

**Comment by merlimat** [comment]
Merged in [branch-1.16](/yahoo/pulsar/tree/branch-1.16) at 956430d57c09d958092a9406b28222057600ae50

**Pull request merged**